src/main/java/io/axoniq/axonserver/connector/query/impl/QueryChannelImpl.java

```diff
@@ -601,6 +622,37 @@ public void sendAck() {
         );
     }
 
+    private static class QueryInProgress {
+
+        private final CompletableFuture<?> closeHandler;
+        private final QueryHandler.FlowControl flowControl;
+
+        public static QueryInProgress noop() {
+            return new QueryInProgress(() -> { }, NoopFlowControl.INSTANCE);
+        }
+
+        public QueryInProgress(Runnable closeHandler, QueryHandler.FlowControl flowControl) {
+            this.closeHandler = new CompletableFuture<>();
+            this.flowControl = flowControl;
+            this.closeHandler.whenComplete((r, e) -> {
+                flowControl.complete();
+                closeHandler.run();
+            });
+        }
+
+        public CompletableFuture<?> completeAsync() {
+            return closeHandler;
+        }
+
+        public void complete() {
```

Comment:
Suggest: rename to `cancel()`

Instead of completing normally, perhaps complete with a `QueryCancelledException` instead?

Reply:
Not sure whether we want to complete exceptionally here. Having a receiver cancelling the stream doesn't seem like an exceptional case but rather a regular one.